### DISCUSSION TITLE
Why do atoms with generic value types produces values of type Awaited<T>?

### DISCUSSION THREAD

**Discussion by elliotwaite** [pmndrs/jotai#1229] [Q&A] [answered]
Given this line:
`const value = useAtomValue(anAtom)`

...if `anAtom` is of type `Atom<T>` (where `T` is a generic type), then `value` will be of type `Awaited<T>` (instead of `T`, which is what I would have expected). Why is that?

[Here's a demo.](https://codesandbox.io/s/jotai-awaited-type-test-unjtdd?file=/src/App.tsx)

And here's the demo's code with comments that explain the issue:
```
type KEY = "a" | "b" | "c";

const NAMES: Record<KEY, string> = {
  a: "a name",
  b: "b name",
  c: "c name"
} as const;

const myKeyAtom = atom<KEY>("a");

function Component<KeyType extends string>({
  keyAtom,
  names
}: {
  keyAtom: Atom<KeyType>;
  names: Record<KeyType, string>;
}) {
  // Here the type of 'key' is 'Awaited<KeyType>'.
  const key = useAtomValue(keyAtom);

  // This code raises the following type error:
  //   Type 'Awaited<KeyType>' cannot be used to index type
  //   'Record<KeyType, string>'.
  const name1 = names[key];

  // However, this works (with the type assertion).
  const name2 = names[key as KeyType];

  return (
    <div>
      {name1}, {name2}
    </div>
  );
}

export default function App() {
  return <Component keyAtom={myKeyAtom} names={NAMES} />;
}

```
Is it expected to have to use type assertions like `key as KeyType` when using atoms with generic types, or am I doing something wrong?

**Comment by dai-shi** [comment, marked as answer]
First of all, typing in jotai is a little bit tricky. It works pretty well for concrete types, but not very well for generic types. for example: #997 
I did my best at the moment, but don't really like some hacky ways like this:
https://github.com/pmndrs/jotai/blob/492544e1845bf2f6d4152242bc64f8dd7d0ccdc0/src/core/atom.ts#L3-L7
I wish some TS wizards help us.

---

So, as you say, `useAtomValue(Atom<T>)` returns `Awaited<T>`. This is because T can be `Promise<U>`.
```js
const anAtom = atom(Promise.resolve("str")); // 👉 Atom<Promise<string>>
const value = useAtom(anAtom); // 👉 string (not Promise<string>)
```

In your case,
```ts
function Component<KeyType extends string>({
```
means `KeyType` **can be** `string | Promise<string>`.
So, `useAtomValue(keyAtom)` should return `Awaited<KeyType>`
If you are sure `KeyType` never becomes `Promise` (the compiler doesn't know),
you can do `useAtomValue(keyAtom) as KeyType`.

In summary, if you understand it, you are doing it correctly. Hope it helps.

**Reply by elliotwaite** [reply]
Okay, got it. Thanks for the explanation!
